mani_skill2/utils/download_asset.py

```diff
@@ -0,0 +1,263 @@
+    """
+    sha256_hash = hashlib.sha256()
+    with open(filename, "rb") as f:
+        for byte_block in iter(lambda: f.read(chunk_size), b""):
+            sha256_hash.update(byte_block)
+    return sha256_hash.hexdigest()
+
+
+def download(
+    url,
+    output_dir=ASSET_DIR,
+    target_path: str = None,
+    checksum=None,
+    verbose=True,
+    non_interactive=True,
+):
+    output_dir = Path(output_dir)
+
+    # Create output directory
+    if not output_dir.exists():
+        if non_interactive or prompt_yes_no(f"{output_dir} does not exist. Create?"):
+            output_dir.mkdir(parents=True)
+        else:
+            return
+
+    if target_path is None:
+        target_path = url.split("/")[-1]
+    output_path = output_dir / target_path
+
+    # Clean up existing files
+    if output_path.exists():
+        if non_interactive or prompt_yes_no(f"{output_path} exists. Remove?"):
+            if output_path.is_dir():
+                shutil.rmtree(output_path)
+            else:
+                output_path.unlink()
+        else:
+            if non_interactive or prompt_yes_no("Continue downloading?"):
+                pass
+            else:
+                print(f"Skip existing: {output_path}")
+                return output_path
+
+    # Download files to temporary location
+    try:
+        if verbose:
+            print(f"Downloading {url}")
+        pbar = tqdm(unit="iB", unit_scale=True, unit_divisor=1024)
+
+        def show_progress(blocknum, bs, size):
+            if blocknum == 0:
+                pbar.total = size
+            pbar.update(bs)
+
+        filename, _ = urllib.request.urlretrieve(url, reporthook=show_progress)
+        pbar.close()
+    except URLError as err:
+        print(f"Failed to download {url}")
+        raise err
+
+    # Verify checksum
+    if checksum is not None and checksum != sha256sum(filename):
+        raise IOError(f"Downloaded file's SHA-256 hash does not match record: {url}")
+
+    # Extract or move to output path
+    if url.endswith(".zip"):
+        with zipfile.ZipFile(filename, "r") as zip_ref:
+            if verbose:
+                for file in tqdm(zip_ref.infolist()):
+                    zip_ref.extract(file, output_dir)
+            else:
+                zip_ref.extractall(output_dir)
+    else:
+        shutil.move(filename, output_path)
+
+    # Explicitly delete the temporary file
+    if Path(filename).exists():
+        Path(filename).unlink()
+
+    return output_path
+
+
+def parse_args():
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        "uid",
+        type=str,
+        help="Asset UID. Use 'all' to download all assets.",
+    )
+    parser.add_argument("--quiet", action="store_true", help="Disable verbose output.")
+    parser.add_argument(
+        "--non-interactive", action="store_true", help="Disable prompts."
+    )
+    return parser.parse_args()
+
```

Comment:
Should we add something to skip re-downloading already downloaded assets? might make running things a bit easier without having to check yourself with workloads that don't spin new instances all the time.

Reply:
Skip re-downloading is already implemented (but it requires users to decide).

The old assets are still kept for backward compatibility. The versioning can somehow be implemented like Habitat, but may need some re-designing.